src/Roslyn.Diagnostics.Analyzers/Core/TestExportsShouldNotBeDiscoverable.cs

```diff
@@ -0,0 +1,89 @@
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            RoslynDiagnosticIds.TestExportsShouldNotBeDiscoverableRuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.RoslyDiagnosticsReliability,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: false,
+            description: s_localizableDescription,
+            helpLinkUri: null,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                var exportAttributeV1 = compilationContext.Compilation.GetTypeByMetadataName("System.ComponentModel.Composition.ExportAttribute");
+                var exportAttributeV2 = compilationContext.Compilation.GetTypeByMetadataName("System.Composition.ExportAttribute");
+
+                if (exportAttributeV1 is null && exportAttributeV2 is null)
+                {
+                    // We don't need to check assemblies unless they're referencing MEF, so we're done
+                    return;
+                }
+
+                compilationContext.RegisterSymbolAction(symbolContext =>
+                {
+                    var namedType = (INamedTypeSymbol)symbolContext.Symbol;
+                    var namedTypeAttributes = namedType.GetApplicableAttributes();
+
+                    AnalyzeSymbolForAttribute(ref symbolContext, exportAttributeV1, namedType, namedTypeAttributes);
+                    AnalyzeSymbolForAttribute(ref symbolContext, exportAttributeV2, namedType, namedTypeAttributes);
+                }, SymbolKind.NamedType);
+            });
+        }
+
+        private static void AnalyzeSymbolForAttribute(ref SymbolAnalysisContext context, INamedTypeSymbol exportAttributeOpt, INamedTypeSymbol namedType, IEnumerable<AttributeData> namedTypeAttributes)
+        {
+            if (exportAttributeOpt is null)
+            {
+                return;
+            }
+
+            var exportAttributeApplication = namedTypeAttributes.FirstOrDefault(ad => ad.AttributeClass.DerivesFrom(exportAttributeOpt));
+            if (exportAttributeApplication is null)
+            {
+                return;
+            }
+
+            if (!namedTypeAttributes.Any(ad =>
+                ad.AttributeClass.Name == nameof(PartNotDiscoverableAttribute)
+                && ad.AttributeClass.ContainingNamespace.Equals(exportAttributeOpt.ContainingNamespace)))
```

Comment:
Attribute might be defined without a ContainingNamespace, null value of `ContainingNamespace` should be handled.

Reply:
If someone adds `global::PartNotDiscoverableAttribute` to Roslyn,sln, they deserve this.

Reply:
But I'll add the null check anyway